scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala

```diff
@@ -51,9 +78,131 @@ object BeamOp extends Serializable {
+                elem.getValue.asScala.map(x => KV.of(elem.getKey, x))
+              })).setCoder(KvCoder.of(OrderedSerializationCoder(ordK), OrderedSerializationCoder(ordV)))
+              .apply(KVToTuple[K, V])
+          }
+        },
+        kryoCoder)
+    }
+
+    def mapSideAggregator(size: Int, semigroup: Semigroup[V])(implicit kryoCoder: KryoCoder): BeamOp[(K, V)] = {
+      TransformBeamOp[(K, V), (K, V)](
+        op,
+        new PTransform[PCollection[(K, V)], PCollection[(K, V)]]() {
+          override def expand(input: PCollection[(K, V)]): PCollection[(K, V)] = {
+            input
+              .apply(TupleToKV[K, V](kryoCoder, kryoCoder))
+              .apply(ParDo.of(MapSideAggregator[K, V](size, semigroup)))
+              .apply(KVToTuple[K, V])
+          }
+        },
+        kryoCoder)
+    }
+  }
+
+  case class SortGroupedValues[K, V](implicit ordV: Ordering[V], kryoCoder: KryoCoder) extends PTransform[PCollection[KV[K, java.lang.Iterable[V]]], PCollection[KV[K, java.lang.Iterable[V]]]] {
+    override def expand(input: PCollection[KV[K, lang.Iterable[V]]]): PCollection[KV[K, lang.Iterable[V]]] = {
+      input
+        .apply(ParDo.of(MapFn[KV[K, java.lang.Iterable[V]], KV[K, java.lang.Iterable[KV[WrappedComparable[V], Unit]]]]{ elem =>
+          KV.of(elem.getKey, elem.getValue.asScala.map(x => KV.of(WrappedComparable(x, ordV), ())).asJava)
+        })).setCoder(KvCoder.of(kryoCoder, IterableCoder.of(KvCoder.of(WrappedAtomicCoder[V](ordV, OrderedSerializationCoder(ordV)), kryoCoder))))
+        .apply(SortValues.create[K, WrappedComparable[V], Unit](BufferedExternalSorter.options()))
+        .apply(ParDo.of(MapFn[KV[K, java.lang.Iterable[KV[WrappedComparable[V], Unit]]], KV[K, java.lang.Iterable[V]]]{ elem =>
+          KV.of(elem.getKey, elem.getValue.asScala.map(_.getKey.value).asJava)
+        })).setCoder(KvCoder.of(kryoCoder, IterableCoder.of(OrderedSerializationCoder(ordV))))
+    }
+  }
+
+  case class TupleToKV[K, V](kCoder: Coder[K], vCoder: Coder[V]) extends PTransform[PCollection[(K, V)], PCollection[KV[K, V]]] {
+    override def expand(input: PCollection[(K, V)]): PCollection[KV[K, V]] = {
+      input
+        .apply(MapElements.via[(K, V), KV[K, V]](new SimpleFunction[(K, V), KV[K, V]]() {
+          override def apply(input: (K, V)): KV[K, V] = KV.of(input._1, input._2)
+        })).setCoder(KvCoder.of(kCoder, vCoder))
+    }
+  }
+
+  case class KVToTuple[K, V](implicit kryoCoder: KryoCoder) extends PTransform[PCollection[KV[K, V]], PCollection[(K, V)]] {
+    override def expand(input: PCollection[KV[K, V]]): PCollection[(K, V)] = {
+      input
+        .apply(MapElements.via[KV[K, V], (K, V)](new SimpleFunction[KV[K, V], (K, V)]() {
+          override def apply(input: KV[K, V]): (K, V) = (input.getKey, input.getValue)
+        })).setCoder(kryoCoder)
+    }
+  }
+
+  case class WrappedComparable[A](value: A, ord: Ordering[A]) extends Comparable[WrappedComparable[A]] {
```

Comment:
Why do we need this class?

Reply:
I suggested it because my reading of SortValues was that is assumes items are `Comparable` and without using it we are discarding the `Ordering[A]` which might happen to be right for some cases, but it will fail with `reverse` and other use cases.